### PR TITLE
Gitlab connector

### PR DESCRIPTION
**Pull request by Dagg-h**
### Issue number

DATA-963

### Description of fix

Adding Gitlab wrapper.

### CODE REVIEW

File: wherescape/connectors/gitlab/python_gitlab_load_data.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    logging.info(rows)
+
+    # Execute the sql
+    wherescape_instance.push_many_to_target(sql, rows)
+    logging.info(f"Data successfully inserted in to the load table.")
+
+    # Final logging
+    end_time = datetime.now()
+    logging.info(
+        "Time elapsed: %s seconds for notion_load_data"
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_load_data.py]: gitlab

File: wherescape/connectors/jira/__init__.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 """
 Module that takes care of the connection to Notion.
 """
-from wherescape.connectors.jira.jira_wrapper import Jira  # noqa: E402
+from ...connectors.jira.jira_wrapper import Jira  # noqa: E402
 
 __all__ = ("Jira",)
```
Comment (BFJonk) [inline wherescape/connectors/jira/__init__.py]: Is this the interface? I would expect only public methods (but I can be wrong)

File: wherescape/connectors/gitlab/__init__.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+"""
+Module that takes care of the connection to Gitlab.
+"""
+from ...connectors.gitlab.gitlab_wrapper import Gitlab  # noqa: E402
+
+__all__ = ("Gitlab",)
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/__init__.py:6]: I can be wrong, but, the __init__.py is used to expose the interface to the outside world, and the interface is not the complete class, isn't it?

File: wherescape/connectors/gitlab/gitlab_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
+import requests
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/gitlab_wrapper.py:2]: missing module comment

File: wherescape/connectors/gitlab/python_gitlab_create_metadata.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                list(value.keys()),
+                list(value.values()),
+                key,
+            )
+
+
+def gitlab_create_metadata(wherescape_instance, columns, python_types, title):
+    start_time = datetime.now()
+    logging.info(
+        "Start time: %s for Gitlab_load_data" % start_time.strftime("%Y-%m-%d %H:%M:%S")
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_create_metadata.py:53]: The Logging library is laready collecting the time, isn't it?

File: wherescape/connectors/gitlab/python_gitlab_create_metadata.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+)
+from ... import WhereScape
+from ...helper_functions import (
+    prepare_metadata_query,
+    create_column_names,
+    create_display_names,
+)
+
+
+def gitlab_create_metadata_project():
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_create_metadata.py]: missing docstrings

File: wherescape/connectors/gitlab/python_gitlab_create_metadata.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    wherescape_instance = WhereScape()
+    access_token = wherescape_instance.read_parameter("gitlab_access_token")
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_create_metadata.py]: should not be parameters, but extended properties

File: wherescape/connectors/gitlab/python_gitlab_create_metadata.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+
+def gitlab_create_metadata(wherescape_instance, columns, python_types, title):
+    start_time = datetime.now()
+    logging.info(
+        "Start time: %s for Gitlab_load_data" % start_time.strftime("%Y-%m-%d %H:%M:%S")
+    )
+    wherescape_object_id = wherescape_instance.object_key
+    types = []
+    for python_type in python_types:
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_create_metadata.py:57]: maybe afterwards check if the len(type_list) = len(columns)?

File: wherescape/connectors/gitlab/python_gitlab_create_metadata.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            types.append("timestamp")
+        elif python_type == "boolean":
+            types.append("boolean")
+
+    # Get the Gitlab database (definition)
+    logging.info(f'Fetched database definition "{title}" from Gitlab')
+
+    # Prepare the sql
+    column_names = create_column_names(columns)
+    display_names = create_display_names(columns)
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_create_metadata.py:72]: Titlecase?

File: wherescape/connectors/gitlab/python_gitlab_create_metadata.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    sql = prepare_metadata_query(
+        wherescape_object_id,
+        "Gitlab api - " + title,
+        columns=column_names,
+        display_names=display_names,
+        types=types,
+        source_columns=columns,
+    )
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_create_metadata.py:80]: does the gitLab APi also prvide column descriptions?

File: wherescape/connectors/gitlab/python_gitlab_load_data.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+
+def gitlab_load_data(load_type):
+    wherescape_instance = WhereScape()
+    start_time = datetime.now()
+    logging.info(
+        "Start time: %s for gitlab_load_data" % start_time.strftime("%Y-%m-%d %H:%M:%S")
+    )
+
+    apikey = wherescape_instance.read_parameter("gitlab_apikey")
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_load_data.py]: should be extended property

File: wherescape/connectors/gitlab/python_gitlab_load_data.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    columns = create_column_names(columns)
+    columns.append("dss_record_source")
+    columns.append("dss_load_date")
+
+    rows = []
+    for row in values:
+        # row = list(row.tolist())
+        row.append("gitlab api - " + load_type)
+        row.append(start_time)
+        rows.append(row)
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_load_data.py]: line break before?

**Review by BFJonk — COMMENTED** [review]
nice overall! Added some comments

File: wherescape/connectors/jira/__init__.py [hunk 13]
```diff
@@ -1,6 +1,6 @@
 """
 Module that takes care of the connection to Notion.
```
Comment (BFJonk) [inline wherescape/connectors/jira/__init__.py]: JIRA

File: wherescape/helper_functions.py [hunk 14]
```diff
@@ -138,3 +160,22 @@ def flatten(x, name=""):
...
+    Parameters:
+    cleaned_json (object): Dict with the flattened and cleaned json response
+    keys_to_keep (dict array): A view object with a list of the keys from a dict
+
+    Returns:
+    out: The dict with all keys, value is None when there was nothing returned from the API
+    """
+    out = {}
+    for key in keys_to_keep:
+        if not key in set(cleaned_json.keys()):
```
Comment (BFJonk) [inline wherescape/helper_functions.py]: test for membership should be 'not in' :
  `if key not in set(cleaned_json.keys()):`

File: wherescape/helper_functions.py [hunk 15]
```diff
@@ -117,18 +128,29 @@ def filter_dict(dict_to_filter, keys_to_keep):
...
         if type(x) is dict:
             for a in x:
-                if name_to_skip == a:
+                if name_to_skip and name_to_skip == a:
                     new_name = name
                 else:
                     new_name = name + a + "_"
                 flatten(x[a], new_name)
-        elif type(x) is list:
+        elif type(x) is list and x:
```
Comment (BFJonk) [inline wherescape/helper_functions.py]: not sure, but if `not x` will `type(x)` than work? in that case `elif x and isinstance(x, list)` would be more logical?

File: wherescape/connectors/gitlab/readme.md [hunk 16]
```diff
@@ -0,0 +1,54 @@
+# Gitlab connector
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/readme.md:1]: suggest to use markdownlint to lint this file (getting 7 warnings)

File: wherescape/connectors/gitlab/readme.md [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+* load_gitlab_project
+* load_gitlab_merge_request
+* load_gitlab_pipeline
+* load_gitlab_tag
+
+Attaching the right scripts is all thats needed to get it working; no input
+in the Source tab is needed.
+
+## Host script
+Create the following new python host script:
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/readme.md]: scripts

File: wherescape/connectors/gitlab/readme.md [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+* python_gitlab_load_data_smart
+* python_gitlab_next_high_water_mark
+* python_gitlab_update_high_water_mark
+
+The scripts should utilize the 4 functions with the same name from the python_ files. Example:
+
+```
+from wherescape.connectors.gitlab.python_gitlab_create_metadata import gitlab_create_metadata_smart
+
+if __name__ == "__main__":
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/readme.md]: why is this line necessary?

File: wherescape/connectors/gitlab/readme.md [hunk 19]
```diff
@@ -0,0 +1,54 @@
+# Gitlab connector
+
+Gitlab connector for WhereScape. Takes care of reading projects, issues, tags,
+pipelines and merge requests from a Gitlab instance.
+
+# Preparation
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/readme.md:6]: These preparation steps could also be automated (or partially)

File: wherescape/connectors/gitlab/readme.md [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+
+## WhereScape Parameters
+
+Add the following parameters to WhereScape:
+* `gitlab_base_url`
+* `gitlab_access_token`
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/readme.md:12]: should preferably be extended properties or properties of a connection  (Wherescape can then take care of encrypting the credentials in the database and on screen)

File: wherescape/connectors/gitlab/readme.md [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+
+if __name__ == "__main__":
+    gitlab_create_metadata_smart()
+```
+
+# Usage
+
+First attach the metadata host script to the load table. After creating the
+table, attach the load_data_smart host script to the load table and plan the
+job in the scheduler. The incremental function should be active after the
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/readme.md]: From this read.me it is not clear what the `incremental function` is or refers to....

File: wherescape/connectors/gitlab/python_gitlab_load_data.py [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+    if "project" in table_name:
+        title = "projects"
+    elif "tag" in table_name:
+        title = "tags"
+    elif "issue" in table_name:
+        title = "issues"
+    elif "pipeline" in table_name:
+        title = "pipelines"
+    elif "merge_request" in table_name:
+        title = "merge_requests"
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_load_data.py]: why not call this `load_type` in stead of `title` ?

File: wherescape/connectors/gitlab/python_gitlab_load_data.py [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+    )
+
+    access_token = wherescape_instance.read_parameter("gitlab_access_token")
+    base_url = wherescape_instance.read_parameter("gitlab_base_url")
+    table_name = wherescape_instance.load_full_name
+
+    since = wherescape_instance.read_parameter("gitlab_high_water_mark")
+    gitlab_instance = Gitlab(access_token, base_url, since)
+
+    if load_type == "projects":
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_load_data.py:53]: I think I would have prefered only one such case statement (in `gitlab_load_data_smart()`). You could have `gitlab_load_data_smart()` return not only the load_type, but also the function name to call. 

Alternatively, you can just leave the load_type out of the input parameter of this function and case on the load_full_name.

File: wherescape/connectors/gitlab/python_gitlab_load_data.py [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+        columns = COLUMN_NAMES_AND_DATA_TYPES["tags"]
+        values = gitlab_instance.get_release_tags()
+    elif load_type == "merge_requests":
+        columns = COLUMN_NAMES_AND_DATA_TYPES["merge_requests"]
+        values = gitlab_instance.get_merge_requests()
+    else:
+        raise Exception("Wrong gitlab load type supplied")
+
+    if len(values) > 0:
+        columns = create_column_names(columns)
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_load_data.py:72]: I don't like the function name `create_column_names`, I think it should be more descriptive and it should be visible that its a wherescape helper function and not a gitlab related function. Suggesting `ws_helper_conform_column_names_to_wherescape` or `conform_column_names_to_wherescape_format`

File: wherescape/connectors/gitlab/python_gitlab_load_data.py [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+        columns.append("dss_record_source")
+        columns.append("dss_load_date")
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_load_data.py:74]: I think we should have a helper function to add dss columns for a certain table type

File: wherescape/connectors/gitlab/python_gitlab_create_metadata.py [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+    them to the WhereScape repository. This is the glue between the
+    gitlab_wrapper and WhereScape.
+    """
+    start_time = datetime.now()
+    logging.info(
+        "Start time: %s for Gitlab_load_data" % start_time.strftime("%Y-%m-%d %H:%M:%S")
+    )
+    wherescape_object_id = wherescape_instance.object_key
+    types = []
+    for python_type in python_types:
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_create_metadata.py:57]: the python to postgresql type-mapping is a good candidate for a helper function. We could add a dict in the helper functions like we did in https://gitlab.wearespindle.com/data/wherescape-warehouse/-/blob/41407c1d5499711d600ef73d70a598c2f49eab34/python/freshservice_api_wrapper.py#L67

File: wherescape/connectors/gitlab/gitlab_data_types_column_names.py [hunk 27]
```diff
@@ -0,0 +1,143 @@
+"""
+COLUMN_NAMES_AND_DATA_TYPES is a dictionary with the flattened values and
+belonging data types returned from the Gitlab API
+"""
+
+COLUMN_NAMES_AND_DATA_TYPES = {
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/gitlab_data_types_column_names.py:6]: Suggest to not only provide data types, but also column descriptions (descriptions column of [https://docs.gitlab.com/ee/api/projects.html#list-all-projects](https://docs.gitlab.com/ee/api/projects.html#list-all-projects) )

File: wherescape/connectors/gitlab/gitlab_wrapper.py [hunk 28]
```diff
@@ -0,0 +1,198 @@
...
+            "per_page": string or int,
+            "current_page": string or int
+        }
+        simple (boolean): If the response of Gitlab should be simplified this needs to be set on True
+        since (string): ISO formatted datetime string to indicate since which date you want values back
+
+        Returns:
+        Formatted url which can be used to make the request
+        """
+        updated_since = f"&updated_after={since}" if since else ""
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/gitlab_wrapper.py:57]: new construct for me.... not sure if I liked it

File: wherescape/connectors/gitlab/gitlab_wrapper.py [hunk 29]
```diff
@@ -0,0 +1,198 @@
...
+        """
+        headers = {
+            "Accept": "application/json",
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {self.access_token}",
+        }
+        response = requests.request(method, url, data=payload, headers=headers)
+        return response
+
+    def format_url(self, resource_api, page_variables, simple, order_by, since):
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/gitlab_wrapper.py]: `since=None` to indicate that since can be optional?

File: wherescape/connectors/gitlab/gitlab_wrapper.py [hunk 30]
```diff
@@ -0,0 +1,198 @@
...
+        """Format URL
+
+        Parameters:
+        resource_api (string): The location of the resource requested
+        page_variables (object): {
+            "per_page": string or int,
+            "current_page": string or int
+        }
+        simple (boolean): If the response of Gitlab should be simplified this needs to be set on True
+        since (string): ISO formatted datetime string to indicate since which date you want values back
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/gitlab_wrapper.py]: doucment that since can be optional

File: wherescape/connectors/gitlab/gitlab_wrapper.py [hunk 31]
```diff
@@ -0,0 +1,198 @@
...
+            response = self.make_request(url, "GET")
+
+            if response.status_code == 403:
+                logging.info(
+                    f"{url}\n Forbidden resource. If you need this resource, please check the user's rights"
+                )
+                current_page = current_page + 1
+                continue
+
+            response.raise_for_status()
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/gitlab_wrapper.py:107]: nice and clean

File: wherescape/connectors/gitlab/gitlab_wrapper.py [hunk 32]
```diff
@@ -0,0 +1,198 @@
...
+    def get_projects_from_api(self):
+        keys_to_keep = COLUMN_NAMES_AND_DATA_TYPES["projects"].keys()
+        resource_api = "projects"
+
+        all_projects = self.paginate_through_resource(
+            resource_api, keys_to_keep, simple="true"
+        )
+        return all_projects
+
+    def get_release_tags(self):
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/gitlab_wrapper.py:146]: I would expect this to be called`get_release_tags_from_api`

File: wherescape/connectors/gitlab/gitlab_wrapper.py [hunk 33]
```diff
@@ -0,0 +1,198 @@
...
+            try:
+                total_pages = response.headers["X-Total-Pages"]
+                current_page = response.headers["X-Page"]
+            except:
+                current_page = current_page + 1
+
+        return all_resources
+
+    def get_projects(self):
+        return self.projects
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/gitlab_wrapper.py:130]: missing docstrings (everywhere)

File: wherescape/connectors/gitlab/gitlab_wrapper.py (hunk 32 above)
Comment (BFJonk) [reply wherescape/connectors/gitlab/gitlab_wrapper.py:146]: also for other functions

File: wherescape/connectors/gitlab/gitlab_wrapper.py [hunk 34]
```diff
@@ -0,0 +1,198 @@
...
+        # projects is a list of tuples, so the first item in the tuple is the id
+        for project in self.projects:
+            project_id = project[0]
+            resource_api = f"projects/{project_id}/merge_requests"
+            project_merge_requests = self.paginate_through_resource(
+                resource_api, keys_to_keep, since=self.since, order_by="title"
+            )
+            all_merge_requests.extend(project_merge_requests)
+
+        return all_merge_requests
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/gitlab_wrapper.py:233]: I would have expected an `if __name__ == "__main__":` line which can be used to test this function without wherescape

File: wherescape/connectors/gitlab/readme.md [hunk 35]
```diff
@@ -0,0 +1,54 @@
...
+
+# Preparation
+
+## WhereScape Parameters
+
+Add the following parameters to WhereScape:
+* `gitlab_base_url`
+* `gitlab_access_token`
+* `gitlab_high_water_mark`
+* `gitlab_high_water_mark_next`
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/readme.md:14]: These do not conform to the standard naming for HWMs . also, the set function should be called from the data-store table

File: wherescape/connectors/gitlab/readme.md [hunk 36]
```diff
@@ -0,0 +1,54 @@
...
+Attaching the right scripts is all thats needed to get it working; no input
+in the Source tab is needed.
+
+## Host script
+Create the following new python host script:
+
+* python_gitlab_create_metadata_smart
+* python_gitlab_load_data_smart
+* python_gitlab_next_high_water_mark
+* python_gitlab_update_high_water_mark
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/readme.md:38]: It's not clear where these host functions should be called

File: wherescape/connectors/gitlab/python_gitlab_high_water_mark.py [hunk 37]
```diff
@@ -0,0 +1,41 @@
+from datetime import datetime
+import logging
+
+from ... import WhereScape
+
+
+def gitlab_next_high_water_mark():
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_high_water_mark.py:7]: it's not clear where / hwne this function is called.

File: wherescape/connectors/gitlab/python_gitlab_high_water_mark.py [hunk 38]
```diff
@@ -0,0 +1,41 @@
...
+    logging.info(f"Next high water mark will be {next_high_water_mark}")
+    wherescape_instance.main_message = (
+        f"Next high water mark will be {next_high_water_mark}"
+    )
+    wherescape_instance.write_parameter(
+        "gitlab_high_water_mark_next", next_high_water_mark
+    )
+
+
+def gitlab_update_high_water_mark():
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_high_water_mark.py:30]: It seems to do something to the lines of [spindle_postgres_pscript_post](https://gitlab.wearespindle.com/data/wherescape-warehouse/-/blob/c4b047e8acad7ea65e635a5a3c2143365046e051/pebble/spindle_postgres_pscript_post.peb) but then differently?

File: wherescape/connectors/gitlab/python_gitlab_high_water_mark.py [hunk 39]
```diff
@@ -0,0 +1,41 @@
...
+    beginning of the job.
+    """
+    wherescape_instance = WhereScape()
+
+    next_high_water_mark = wherescape_instance.read_parameter(
+        "gitlab_high_water_mark_next"
+    )
+    wherescape_instance.write_parameter("gitlab_high_water_mark", next_high_water_mark)
+    wherescape_instance.main_message = (
+        f"High water mark is set to {next_high_water_mark}"
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_high_water_mark.py:43]: I would include a timestamp, saying when the update was done see spindle_postgres_pscript_post

File: wherescape/connectors/gitlab/python_gitlab_high_water_mark.py [hunk 40]
```diff
@@ -0,0 +1,41 @@
...
+        f"Next high water mark will be {next_high_water_mark}"
+    )
+    wherescape_instance.write_parameter(
+        "gitlab_high_water_mark_next", next_high_water_mark
+    )
+
+
+def gitlab_update_high_water_mark():
+    """
+    After all objects have been successfull loaded, the high water mark is
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/python_gitlab_high_water_mark.py:32]: hight water mark should be set after datastore, not after loading

**Review by BFJonk — CHANGES_REQUESTED** [review]
Some general comments: 
- requirements.txt is missing
- previous comments have not been processed
- I think you should use pylint / flake8 / prospector to reduce a number of potential issues like : 
   - too long lines
   - Dangerous default value [] as argument

File: wherescape/connectors/gitlab/gitlab_wrapper.py [hunk 41]
```diff
@@ -0,0 +1,198 @@
...
+        all_resources = []
+
+        while current_page < total_pages:
+            page_variables = {"per_page": per_page, "current_page": current_page}
+            url = self.format_url(resource_api, page_variables, simple, order_by, since)
+
+            response = self.make_request(url, "GET")
+
+            if response.status_code == 403:
+                logging.info(
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/gitlab_wrapper.py:101]: Is this just an info statement? or a should this be a warning?

File: wherescape/connectors/gitlab/gitlab_wrapper.py [hunk 42]
```diff
@@ -0,0 +1,198 @@
...
+                    f"{url}\n Forbidden resource. If you need this resource, please check the user's rights"
```
Comment (BFJonk) [inline wherescape/connectors/gitlab/gitlab_wrapper.py:102]: The info statement seems not to print the URL : 
image

File: wherescape/connectors/gitlab/gitlab_wrapper.py (hunk 41 above)
Comment (Shrikey) [reply wherescape/connectors/gitlab/gitlab_wrapper.py:101]: Thought about it, but info is fine; if there's a `secret` project, it shouldnt be in the warehouse.

File: wherescape/connectors/gitlab/gitlab_wrapper.py (hunk 34 above)
Comment (Shrikey) [reply wherescape/connectors/gitlab/gitlab_wrapper.py:233]: I don't think that is necessary; people can create their own 'test' files or we should create proper tests.

File: wherescape/connectors/gitlab/python_gitlab_create_metadata.py (hunk 10 above)
Comment (Dagg-h) [reply wherescape/connectors/gitlab/python_gitlab_create_metadata.py:80]: Unfortunately not!

File: wherescape/helper_functions.py (hunk 15 above)
Comment (Dagg-h) [reply wherescape/helper_functions.py]: It checks if the list is not empty. For me it makes sense to check first: is it a list, and if so: is the list not empty? Then do something. I will change to and len(x) > 0

File: wherescape/connectors/gitlab/gitlab_data_types_column_names.py (hunk 27 above)
Comment (Dagg-h) [reply wherescape/connectors/gitlab/gitlab_data_types_column_names.py:6]: We will put a link in the the wherescape loadtable descriptions, pointing towards the API docs with the fields.

File: wherescape/connectors/gitlab/gitlab_wrapper.py (hunk 32 above)
Comment (Dagg-h) [reply wherescape/connectors/gitlab/gitlab_wrapper.py:146]: Changed the names around! Made the get_projects to get_projects_loaded to indicate it will get prefetched data.

**Comment by Dagg-h** [comment]
Pull request merged into main.

File: wherescape/connectors/gitlab/__init__.py (hunk 3 above)
Comment (BFJonk) [reply wherescape/connectors/gitlab/__init__.py:6]: I promised to send a reference about this comment: https://www.reddit.com/r/Python/comments/1bbbwk/comment/c95cjs5/?utm_source=share&utm_medium=ios_app&utm_name=iossmf&context=3
